### DISCUSSION TITLE
Conditional alias

### DISCUSSION THREAD

**Discussion by m-lima** [nushell/nushell#7505] [Q&A]
How can a conditional alias be defined?

For instance, it is common to have a dotfiles repository that multiple systems use. And it might be desirable to conditionally define aliases:
```nu
if (which nvim) != null {
  alias vi = nvim
} else if (which vim) != null {
  alias vi = vim
}
```

In this simple example, when the dotfiles are installed in systems without `neovim`, it'll revert to `vim`, or ultimately fallback to no alias an call `vi`

A naïve attempt at something like this would be:
```nu
alias vi = ([(which nvim) (which vim) (which vi)] | flatten | first | get path)
```

But this has two drawbacks:
- It will run this whole pipeline every time the `vi` alias is called, even though we are only interested in the result during the **definition** of the alias
- And, more importantly, this does not work, since it interprets as us trying to output a string

Has anyone tried to create a conditional alias?
Any pointers or suggestions?

I understand that this is achievable by doing something like:
```nu
def vi [...args] {
  if (which nvim) != null {
    nvim $args
  }
  [...other checks and branches for vim and vi...]
}
```

But, again, there are drawbacks:
- Once more it will re-run all the check, even though I'm only interested in the "define-time" evaluation of `which nvim`
- And, more importantly, It breaks when called with flags. E.g.: `vi -O fileA.txt fileB.txt`

**Comment by fdncred** [comment]
I agree that `if` should change to allow a more intuitive functionality. I believe this is the current solution but there appears to be a bug in alias that crashes nushell (#7512)
```
alias vi = if not ((which nvim) | is-empty) { nvim } else if not ((which vim) | is-empty) { vim }
```

**Reply by m-lima** [reply]
Thanks for the suggestion! This is interesting
Though it under performs compared to the custom command exemplified above, as it cannot take any arguments (as opposed to only not able to handle flags)

Let's hope that the `if` gets revamped to allow such things. The recent `mut` variables being manipulated inside blocks (and, by extension, `if` blocks) is a great advance. However, it doesn't apply to aliases :(
